### PR TITLE
Issue 651: add schema validation before QObj jobs are sent

### LINKED ISSUE #651: Introduce validation against the schemas
### What is the expected enhancement?
According to the [robustness principle](https://en.wikipedia.org/wiki/Robustness_principle), Qiskit should validate that the `Qobj` emitted complies with the schemas.

If doing this is something that should be done upon a new request or during testing, is open for debate.

### PR DESCRIPTION
**Pull request by Max-Reuter**
### Summary
Runs validation on `Qobj` objects sent to local and non-local job requests, throwing an error if they don't match the `JSON` schema.

### Details and comments
Fix #651.

### CODE REVIEW

File: qiskit/backends/ibmq/ibmqbackend.py [hunk 1]
```diff
@@ -10,13 +10,17 @@
...
+import json
+import jsonschema
 
 from IBMQuantumExperience import ApiError
 from qiskit import QISKitError
 from qiskit._util import _camel_case_to_snake_case, AvailableToOperationalDict
 from qiskit.backends import BaseBackend
 from qiskit.backends.ibmq.ibmqjob import IBMQJob
 from qiskit.backends import JobStatus
+from qiskit import __path__ as qiskit_path
```
Comment (ewinston) [inline qiskit/backends/ibmq/ibmqbackend.py]: This is probably not necessary since `qiskit.__path__[0]` or `os.path.dirname(qiskit.__file__)` is common and you only use it once.

File: qiskit/backends/ibmq/ibmqbackend.py [hunk 2]
```diff
@@ -51,9 +55,27 @@ def run(self, qobj):
...
         """
+
+        # verify the QObj is valid before making requests for computing resources
+        sdk = qiskit_path[0]
+        # Schemas path:     qiskit/backends/schemas
+        schemas_path = os.path.join(sdk, 'schemas')
+        schema_file_path = os.path.join(schemas_path, 'qobj_schema.json')
+
+        with open(schema_file_path, 'r') as schema_file:
+            schema = json.load(schema_file)
```
Comment (ewinston) [inline qiskit/backends/ibmq/ibmqbackend.py]: Since this only needs to be done once, loading the schema could be moved to `__init__`.

File: qiskit/backends/ibmq/ibmqbackend.py [hunk 3]
```diff
@@ -51,9 +55,27 @@ def run(self, qobj):
...
+
+        try:
+            jsonschema.validate(qobj.as_dict(), schema)
+        except jsonschema.ValidationError as validation_error:
+            raise ValueError(validation_error)
```
Comment (ewinston) [inline qiskit/backends/ibmq/ibmqbackend.py]: I believe within qiskit we are always raising errors derived from qiskit.QISKitError.

File: qiskit/backends/local/localjob.py [hunk 4]
```diff
@@ -55,6 +59,20 @@ def __init__(self, fn, qobj):
...
+        sdk = qiskit_path[0]
+        # Schemas path:     qiskit/backends/schemas
+        schemas_path = os.path.join(sdk, 'schemas')
+        schema_file_path = os.path.join(schemas_path, 'qobj_schema.json')
+
+        with open(schema_file_path, 'r') as schema_file:
+            schema = json.load(schema_file)
+
+        try:
+            jsonschema.validate(self._qobj.as_dict(), schema)
```
Comment (ewinston) [inline qiskit/backends/local/localjob.py]: It seems like the validation should happen in the same class object for both `ibmq` and `local` backends. Now it seems to be happening in `ibmqbackend` and `localjob`.

Comment (ewinston) [reply qiskit/backends/local/localjob.py]: Perhaps it makes sense in BaseJob derived classes so in ibmqjob.py validation could be done in the `else` clause of this statement;
https://github.com/Qiskit/qiskit-terra/blob/d11bdd2523aba4b8f49d4fc8cf927a288b19cf3b/qiskit/backends/ibmq/ibmqjob.py#L139-L148

File: test/python/test_qobj.py [hunk 5]
```diff
@@ -87,3 +90,21 @@ def test_from_dict_per_class(self):
...
+        bad_qobj = self.valid_qobj
+        bad_qobj.experiments = None  # set experiments to None to cause the qobj to be invalid
+
+        local_backend = FakeBackend()
+        bad_qobj.header = QobjHeader(backend_name=local_backend.name)
+        # test access denial against a local backend
+        with self.assertRaises(ValueError):
+            localjob.LocalJob(None, bad_qobj)
+
+        config = DummySimulator.DEFAULT_CONFIGURATION
```
Comment (delapuente) [inline test/python/test_qobj.py]: Separate the tests into two: one for local jobs, the other for IBM Q jobs.

File: test/python/test_qobj.py [hunk 6]
```diff
@@ -87,3 +90,21 @@ def test_from_dict_per_class(self):
         for qobj_class, (qobj, expected_dict) in test_parameters.items():
             with self.subTest(msg=str(qobj_class)):
                 self.assertEqual(qobj, qobj_class.from_dict(expected_dict))
+
+    def test_ibmqobj_raises_when_sending_bad_qobj(self):
+        """Test IBMQobj is denied resource request access when invalid."""
```
Comment (delapuente) [inline test/python/test_qobj.py]: Please, clarify the docstring by replacing `invalid` with `invalid Qobj instance`.

File: test/python/test_qobj.py [hunk 7]
```diff
@@ -87,3 +90,21 @@ def test_from_dict_per_class(self):
...
+        bad_qobj = self.valid_qobj
+        bad_qobj.experiments = None  # set experiments to None to cause the qobj to be invalid
+
+        local_backend = FakeBackend()
+        bad_qobj.header = QobjHeader(backend_name=local_backend.name)
+        # test access denial against a local backend
```
Comment (delapuente) [inline test/python/test_qobj.py]: Remove the comment.

File: qiskit/backends/ibmq/ibmqbackend.py (hunk 1 above)
Comment (Max-Reuter) [reply qiskit/backends/ibmq/ibmqbackend.py]: Shall I `import qiskit` then, so I can access `qiskit.__path__`?

Comment (ewinston) [reply qiskit/backends/ibmq/ibmqbackend.py]: I think that sounds fine.

File: qiskit/backends/local/localjob.py (hunk 4 above)
Comment (ewinston) [reply qiskit/backends/local/localjob.py]: I think yours is the better suggestion.

Comment (ewinston) [reply qiskit/backends/local/localjob.py]: We are still bringing up a device backend which can understand qobj 1.0. I'm not sure if it should be done then or when all devices have been updated.

**Comment by ajavadia** [comment]
Separating loading and validation function sounds like the right thing to do. I think Thomas already did something like this when he was validating the examples in the spec. @taalexander, can you please see if your code is applicable here?

**Comment by taalexander** [comment]
@ajavadia what I have should work here. I'll submit a PR against @Max-Reuter's branch.

**Comment by taalexander** [comment]
@delapuente this sounds good to me. It would be good to write the validation functions generally so that these routines may be used for all of the schemas in `qiskit/schemas`. 
- Functions to load and create a `jsonschema.IValidator` (ie. `_validator.py`) should be placed not in  `qiskit/qobj/_validation.py` but in `qiskit/_validation.py` as other schemas in `qiskit/schemas/` will require these functions. A validation function may then be created in `qiskit/qobj/_validation.py`. 
- Rather than raising`QiskitError`, the error `qobj._utils.QobjValidationError` may be raised, although this is just a trivial subclass of `QiskitError`. I might consider creating a more general error such as `JSONValidationError` inherting from `QiskitError` when looking forward to validating the other JSON schemas. 
- Tests should be added which run the validation routines against relevant examples in`qiskit/schemas/examples` which would aim at incorporating `qiskit-terra/test/schemas/check_schemas.py` into the testing framework.

**Comment by delapuente** [comment]
Thank you, @taalexander. I think testing the examples against the proper schemas is a very good idea but I prefer to have a separate issue for that and get this out as soon as possible. Do you have a PR ready for that or are you tracking that idea in an already open issue?

**Comment by delapuente** [comment]
The PR is ready. The following tasks are meant to be implemented in several follow-ups:

It remains:
1. Make Qiskit unrollers to produce Qobj conformant outputs.
2. When sending the Qobj to the backends, after converting to
the old format, restore conditionals using the new `bfunc` to
the old format using the `conditional` field.
3. Implement an extension mechanism for schemas allowing
backends to alter the base Qobj schema.
4. Remove the transitional Qobj schema and make Qiskit validate
against "Qobj base schema" + "backend-specific schema".
5. Improve `_validation.py` utilities for dealing with validation errors.

**Comment by delapuente** [comment]
@diego-plan9 would you mind to take a look?

File: qiskit/backends/ibmq/ibmqjob.py [hunk 8]
```diff
@@ -129,14 +130,20 @@ def __init__(self, api, is_device, qobj=None, job_id=None, backend_name=None,
             backend_name(String): The name of the backend that run the job.
             creation_date(String): When the job was run.
 
+        Raises:
+            QISKitError: if the input Qobj is invalid
```
Comment (diego-plan9) [inline qiskit/backends/ibmq/ibmqjob.py]: This one is probably a `QobjValidationError`? For consistency, can you also either add a similar `Raises` in the docstring of the corresponding function of `LocalJob` - or remove both of them, as we don't really enforce declaring the errors that are raised in children functions (maybe we should at some point and in some cases such as this one, though, as it can be quite handy).

File: qiskit/qobj/_validation.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        jsonschema.validate(qobj.as_dict(), qobj_schema)
+    except jsonschema.ValidationError as err:
+        newerr = QobjValidationError(
+            'Qobj failed validation. Set Qiskit log level to DEBUG for further information.')
+        newerr.__cause__ = None
+        logger.debug('%s', _format_causes(err))
+        raise newerr
+
+
+def _format_causes(err, level=0):
```
Comment (diego-plan9) [inline qiskit/qobj/_validation.py]: Can you add a docstring to this function?

File: qiskit/qobj/_validation.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    """Validates a QObj against a schema."""
+    qobj_schema = _load_qobj_schema()
+
+    try:
+        jsonschema.validate(qobj.as_dict(), qobj_schema)
+    except jsonschema.ValidationError as err:
+        newerr = QobjValidationError(
+            'Qobj failed validation. Set Qiskit log level to DEBUG for further information.')
+        newerr.__cause__ = None
+        logger.debug('%s', _format_causes(err))
```
Comment (diego-plan9) [inline qiskit/qobj/_validation.py]: Would it be possible to store the `err` somewhere in `newerr`, instead of discarding the information altogether? This would allow real debugging and accessing that information regardless of the log settings.

File: qiskit/qobj/_validation.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+_SCHEMAS = {}
+
+
+logger = logging.getLogger(__name__)
+
+
+def _load_qobj_schema():
+    """Loads the QObj schema for use in future validations."""
+    if 'qobj' not in _SCHEMAS:
```
Comment (diego-plan9) [inline qiskit/qobj/_validation.py]: Would it make sense to make this function accept a parameter, and load a specific schema based on that parameter, instead of hard-coding the dict key? It feels a bit counter-intuitive to have the function load "all" the schemas (even if "all" currently means 1) and populate the dict unconditionally, making it a bit less dynamic than it will be needed as soon as more than one schema needs to be loaded. Might be something that is better tackled in future iterations?

File: qiskit/unroll/_jsonbackend.py [hunk 12]
```diff
@@ -254,14 +254,16 @@ def drop_condition(self):
...
+    def start_gate(self, name, args, qubits, nested_scope=None, extra_fields=None):
         """Begin a custom gate.
 
         name is name string.
         args is list of Node expression objects.
         qubits is list of (regname, idx) tuples.
         nested_scope is a list of dictionaries mapping expression variables
         to Node expression objects in order of increasing nesting depth.
+        extra_fields is a dictionary allowing the extension or overriding of
+        the gate instruction properties.
```
Comment (diego-plan9) [inline qiskit/unroll/_jsonbackend.py]: Can you take the chance to change this into a proper docstring, even if technically not part of the PR?

File: qiskit/qobj/_validation.py (hunk 11 above)
Comment (taalexander) [reply qiskit/qobj/_validation.py]: I believe this method should be added to a module in the directory above `qiskit/_validation.py`. It will be necessary for loading of all schemas in `qiskit/schemas`. I agree with @diego-plan9 it should accept a parameter specifying the schema rather than hardcoding the schema.

File: qiskit/qobj/_validation.py [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+        # valid Qobj structures and the Json Schema extension mechanism is in
+        # place.
+        schema_file_path = os.path.join(sdk, 'schemas', 'transitional_qobj_schema.json')
+        with open(schema_file_path, 'r') as schema_file:
+            _SCHEMAS['qobj'] = json.load(schema_file)
+
+    return _SCHEMAS['qobj']
+
+
+def validate_qobj_against_schema(qobj):
```
Comment (taalexander) [inline qiskit/qobj/_validation.py]: This method and the functionality of the error it throws is quite generic to the validation of any JSON schema for qiskit. I believe the method should be moved up to `qiskit/_validation.py`. The method should accept a schema as an argument. and possibly be renamed something like `validate_json_schema(json,schema)`.

File: qiskit/qobj/_validation.py [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+    """Validates a QObj against a schema."""
+    qobj_schema = _load_qobj_schema()
+
+    try:
+        jsonschema.validate(qobj.as_dict(), qobj_schema)
```
Comment (taalexander) [inline qiskit/qobj/_validation.py]: By default `jsonschema.validate` will validate the schema. It would be better to create a new method 
```
def create_validator(schema, check_schema=True,
                     validator_class=None, *args, **kwargs):
    """
    Generate validator for JSON schema.

    Parameters
    ----------
    schema (dict): JSON schema `dict`.

    check_schema (bool): Verify schema is valid.

    validator_class (jsonschema.IValidator): JsonSchema IValidator instance. Default
            behavior is to determine this from the schema `$schema` field.

    Other positional and keywork arguments will be passed onto `jsonschema`
    validator.
    Returns
    -------
    validator (qutip.IValidator): Validator for JSON schema.
    """



    if not isinstance(schema, dict):
        raise ValueError('''Supplied schema must be a Python dictionary
                            with proper schema form.
                         ''')
    #Resolve Json spec from schema if needed
    if validator_class is None:
        validator_class = jsch.validators.validator_for(schema)

    #verify supplied schema is valid
    if check_schema:
        validator_class.check_schema(schema)

    # generate and return qobj validator
    return validator_class(schema, *args, **kwargs)
```

This will return a validator that may be called as `validator.validate`. This also had the nice feature that we may call `create_validator(...,check_schema=True) in our test suite against all schemas to verify that they are compliant. This work would then not need to be repeated for every validation call.

File: qiskit/qobj/_validation.py [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+    except jsonschema.ValidationError as err:
+        newerr = QobjValidationError(
```
Comment (taalexander) [inline qiskit/qobj/_validation.py]: Would it be possible to make this error inherit from a more generic error along the lines of `QiskitJSONValidationError` so that the same error type could be reused for validation of other JSON schemas in qiskit?

File: qiskit/backends/ibmq/ibmqjob.py (hunk 8 above)
Comment (delapuente) [reply qiskit/backends/ibmq/ibmqjob.py]: I'm going to remove them since they should be raised in the `submit` functions but leave the proper notes to do it.

File: qiskit/qobj/_validation.py (hunk 11 above)
Comment (delapuente) [reply qiskit/qobj/_validation.py]: I'm aware but did not want to generalize yet. I'm not sure if this should be a parametrized version of the function or a singleton object. Neither I'm sure if the dictionary should be indexed by name or by the schema path and linked to each Qobj class with an inner field. I think my preferred syntax would be something like:

```python
from schema import get_validator
get_validator().validate(qobj)
get_validator().validate(result)
get_validator().validate(backend_configuration)
```

And internally:

```python
# module schema
def get_validator():
  return _VALIDATOR_SINGLETON

class Validator():
  def validate(self, obj):
    schema = self._get_schema(obj.schema)
    jsonvalidator.validate(serialize_to_json(obj), schema)
```

File: qiskit/qobj/_validation.py (hunk 10 above)
Comment (delapuente) [reply qiskit/qobj/_validation.py]: It is automatically in `__cause__` and I'm removing it intentionally since the output is noisy and almost useless but I can restore the original behaviour.

**Comment by diego-plan9** [comment]
Thanks @Max-Reuter and @delapuente for a nifty tackling of the issue, which spawned in a number of ramifications gracefully handled! Please note that it's important to complete the plan sketched by @delapuente  at https://github.com/Qiskit/qiskit-terra/pull/807#issuecomment-417040576 after 0.6, to ensure the schema and the Qobj definition remains close to the original intent.

**Comment by delapuente** [comment]
The plan has been outlined in #851
